model/LtiLaunchDataService.php

```diff
@@ -40,16 +43,30 @@ class LtiLaunchDataService extends ConfigurableService
      */
     public function findDeliveryFromLaunchData(LtiLaunchData $launchData)
     {
-        $deliveryUri = $launchData->getCustomParameter('delivery');
-
-        if (!is_null($deliveryUri)) {
+        return $this->findResource($launchData->getCustomParameter('delivery'));
+    }
 
-            $delivery = new \core_kernel_classes_Resource($deliveryUri);
-        } else {
-            // stored in link
-            $delivery = LTIDeliveryTool::singleton()->getDeliveryFromLink();
-        }
+    /**
+     * @param LtiLaunchData $launchData
+     *
+     * @return core_kernel_classes_Container|core_kernel_classes_Resource
+     * @throws common_exception_Error
+     */
+    public function findDeliveryExecutionFromLaunchData(LtiLaunchData $launchData)
+    {
+        return $this->findResource($launchData->getCustomParameter('execution'));
+    }
 
-        return $delivery;
+    /**
+     * @param $uri
+     *
+     * @return core_kernel_classes_Container|core_kernel_classes_Resource
+     * @throws common_exception_Error
+     */
+    private function findResource($uri)
+    {
+        return $uri !== null
+            ? new core_kernel_classes_Resource($uri)
```

Comment:
Could you please add getting the resource using OntologyAwareTrait ($this->getResource($uri))